src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java

```diff
@@ -0,0 +1,224 @@
+     * @param scvh Server Response Version Header name
+     * @param cmch Client Modification Check Header name
+     * @param wire Original wire
+     */
+    AbstractHeaderBasedCachingWire(
+            final String scvh,
+            final String cmch,
+            final Wire wire
+    ) {
+        this.scvh = scvh;
+        this.cmch = cmch;
+        this.origin = wire;
+        this.cache = new ConcurrentHashMap<>();
+    }
+
+    // @checkstyle ParameterNumber (7 lines)
+    @Override
+    public final Response send(
+            final Request req,
+            final String home,
+            final String method,
+            final Collection<Map.Entry<String, String>> headers,
+            final InputStream content,
+            final int connect,
+            final int read
+    ) throws IOException {
+        final Response rsp;
+        if (method.equals(Request.GET)) {
+            rsp = this.consultCache(
+                    req, home, method, headers, content, connect, read
+            );
+        } else {
+            rsp = this.origin.send(
+                    req, home, method, headers, content, connect, read
+            );
+        }
+        return rsp;
+    }
+
+    /**
+     * Check cache and update if needed.
+     *
+     * @param req Request
+     * @param home URI to fetch
+     * @param method HTTP method
+     * @param headers Headers
+     * @param content HTTP body
+     * @param connect The connect timeout
+     * @param read The read timeout
+     * @return Response obtained
+     * @throws IOException if fails
+     * @checkstyle ParameterNumber (6 lines)
+     */
+    private Response consultCache(
+        final Request req,
+        final String home,
+        final String method,
+        final Collection<Map.Entry<String, String>> headers,
+        final InputStream content,
+        final int connect,
+        final int read
+    ) throws IOException {
+        final Response rsp;
+        if (this.cache.containsKey(req)) {
+            rsp = this.validateCacheWithServer(
+                    req, home, method, headers, content, connect, read
+            );
+        } else {
+            rsp = this.origin.send(
+                    req, home, method, headers, content, connect, read
+            );
+            this.updateCache(req, rsp);
+        }
+        return rsp;
+    }
+
+    /**
+     * Check response and update cache if needed.
+     * @todo #90:30min Evict cache entry if If-Modified-Since request responded
+     *  with HTTP_OK code and no Last-Modified header.
+     * @param req Request
+     * @param home URI to fetch
+     * @param method HTTP method
+     * @param headers Headers
+     * @param content HTTP body
+     * @param connect The connect timeout
+     * @param read The read timeout
+     * @return Response obtained
+     * @throws IOException if fails
+     * @checkstyle ParameterNumber (8 lines)
+     */
+    private Response validateCacheWithServer(
+        final Request req,
```

Comment:
@gmandnepr please read http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#indentation - these arguments could be compacted to smaller number of lines